Allow mapping of UserSessionNotes into UserInfo

sonOfRa opened the pull request:
Fixes #15369

Simply adds the interface as a marker, which is then picked up by `OIDCAttributeMapperHelper#addIncludeInTokensConfig` to add the configuration option automatically.

sonOfRa committed:
Allow mapping of UserSessionNotes into UserInfo

Fixes #15369

thomasdarimont commented:
This seems to be a duplicate of https://github.com/keycloak/keycloak/pull/15263

mposolda commented:
@thomasdarimont Thanks for pointing that!

@sonOfRa This PR may require test (see https://github.com/keycloak/keycloak/pull/15263 for the details). I am keeping this PR open for now. When test is added, we can merge that one PR, which will first add a test :-) Another hint: `AbstractKerberosSingleRealmTest.credentialDelegationTest()` uses UserSessionNote mapper under the covers. Maybe it helps to just add userInfo request into that test to make sure that claim is returned in userInfo?

sonOfRa commented:
I added a test as an additional test in `AbstractKerberosSingleRealmTest`, thanks for pointing me in the right direction.

That required a bit of fiddling with the SPNEGO helper methods so I could actually get the access token as a string, rather than just the parsed object.

I also added an overload to `UserSessionNoteMapper.createClaimMapper(...)` which retains the old behaviour for the old method call (no adding of claims to the userinfo endpoint by default), and adds a `userInfo` parameter so the behaviour can be toggled.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/kerberos/AbstractKerberosTest.java:
```diff
@@ -209,6 +209,36 @@ protected AccessToken assertSuccessfulSpnegoLogin(String clientId, String loginU
+        return assertSpnegoLoginTokenResponse("kerberos-app", loginUsername, expectedUsername, password);
+    }
+
+    protected OAuthClient.AccessTokenResponse assertSpnegoLoginTokenResponse(String clientId, String loginUsername, String expectedUsername, String password) throws Exception {
```
IS it possible to update existing method `assertSuccessfulSpnegoLogin` to use your newly added method to avoid so much code duplications?

I am thinking about something around:
```
 protected AccessToken assertSuccessfulSpnegoLogin(String clientId, String loginUsername, String expectedUsername, String password) throws Exception {
        OAuthClient.AccessTokenResponse tokenResponse = assertSpnegoLoginTokenResponse(.. arguments of your new method here ...;

        AccessToken token = oauth.verifyToken(tokenResponse.getAccessToken());
        Assert.assertEquals(userId, token.getSubject());
        Assert.assertEquals(expectedUsername, token.getPreferredUsername());
         return token;
}
```

sonOfRa replied:
I got rid of that method entirely. The return value AccessToken is only used in one other test, which I adjusted by adding an addition `oauth.verifyToken(response.getAccessToken())` call there.

This feels much cleaner to me, and with relatively minimal overhead of running an extra token verification in a few cases.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/kerberos/AbstractKerberosSingleRealmTest.java:
```diff
@@ -216,4 +213,53 @@ public void credentialDelegationTest() throws Exception {
     }
+
+    @Test
+    public void credentialDelegationUserInfoTest() throws Exception {
```
Is it possible to use existing method `credentialDelegationTest()` and just add the call to userInfo to that method?

I know that you need to use AccessTokenResponse, but you can possibly just use your new method to retrieve `AccessTokenResponse` and then do simply:
```
AccessToken token = oauth.verifyToken(tokenResponse.getAccessToken());
```
And then you have both `tokenResponse` and `token` though?

sonOfRa replied:
Done, integrated the test into the existing `credentialDelegationTest()`

mposolda reviewed (changes requested):
@sonOfRa Thanks! Overaly looks like you're in good direction. Would be just good to avoid introducing so much code duplicities and instead re-use as much as possible for various tests - added some inline comments regarding that.

mposolda reviewed (approved):
@sonOfRa That works, Thanks! I hope to merge once tests are OK
